### PR TITLE
Add Solana Contract

### PR DESCRIPTION
**Pull request by freddyaboulton**
# Overview

Modifies the Candy Machine program in the following ways:

- Deletes `AddConfigLines` and other methods so that the metadata URI is not stored on chain to save on gas. 
- Adds `token_index` argument to `mint_nft` so that we can control which URI is minted
- Adds a `ClaimStatus` account that keeps track of which NFTs have been claimed to prevent anyone from minting the same NFT twice

Adds an express webserver with the following route:

```
app.post('/claim/:tokenIndex', ...
```

The idea is that the front-end can query metamask/ETH contract to see which `tokenIndex` is owned by the ETH address. The front-end can also read the solana address from the Phantom Wallet. Then the front-end can make a request to the above route to mint that NFT to that address.

# How to test

### Unit Tests
- Contract unit tests are in `pmv-sol`. Run `npm test`. 

### Web server
- Create a default solana wallet by following these [instructions](https://docs.solana.com/wallet-guide/paper-wallet)
- Airdrop some SOL into this wallet: `solana airdrop 5 --url https://api.devnet.solana.com`. Might need more than 5 SOL for this.
- Create a new wallet that will be a stand-in for a user who has bought an eth NFT: ` solana-keygen new -o ~/newkey-2.json`. Note the public key.
- Start the server from the `app` directory in `pmv-sol`: `env MY_WALLET=$HOME/.config/solana/id.json node app.js`
- Run the following command: ` curl -d "solAddress=<sol-address>" -X POST http://localhost:3000/claim/<token-index>`. Make sure `1 <= token-index <=10` so that the metadata properly shows up on the block explorer.
- Verify that the token was minted to the `newkey-2.json` with `spl-token accounts --owner ~/newkey-2.json`
image
- See it on the [block explorer](https://explorer.solana.com/address/Auh2DBYe9qrMcAfyxdJmrBLeSCAQna3uVaCx4JGzD2Lg?cluster=devnet)!
image

### CODE REVIEW

File: pmv-sol/app/app.js [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import express from 'express';
+import { initialize } from './helpers.js';
+import { mintToken } from './claim.js';
+
+const app = express()
+const port = 3000
+
+await initialize();
```
Comment (freddyaboulton) [inline pmv-sol/app/app.js:8]: This initializes the contract everytime the server is launched. Which is good for testing but something we'll move away from later.

File: pmv-sol/migrations/deploy.ts [hunk 2]
```diff
@@ -0,0 +1,12 @@
+// Migrations are an early feature. Currently, they're nothing more than this
```
Comment (freddyaboulton) [inline pmv-sol/migrations/deploy.ts:1]: Auto-generated

File: pmv-sol/Anchor.toml [hunk 3]
```diff
@@ -0,0 +1,12 @@
+[programs.localnet]
```
Comment (freddyaboulton) [inline pmv-sol/Anchor.toml:1]: auto-generated

File: pmv-sol/programs/pmv/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,458 @@
+pub mod utils;
```
Comment (freddyaboulton) [inline pmv-sol/programs/pmv/src/lib.rs:1]: Modified candy machine contract

File: pmv-sol/app/app.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+app.get("/", function (req, res) {
+    res.send('Hello World!');
+})
+
+app.get('/claim/:solAddress/:tokenIndex', function (req, res) {
```
Comment (freddyaboulton) [inline pmv-sol/app/app.js:14]: This should probably be a put?

File: pmv-sol/programs/pmv/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+    anchor_spl::token::Token,
+    metaplex_token_metadata::{
+        instruction::{create_master_edition, create_metadata_accounts, update_metadata_accounts},
+        state::{MAX_CREATOR_LEN, MAX_CREATOR_LIMIT, MAX_SYMBOL_LENGTH},
+    },
+    spl_token::state::Mint,
+};
+anchor_lang::declare_id!("2MvgrbsWoramiYaBBE9pqhXcg72ByZvbMJCNDbrekvHV");
+
+const PREFIX: &str = "candy_machine";
```
Comment (freddyaboulton) [inline pmv-sol/programs/pmv/src/lib.rs:18]: I thought about changing all the references on `candy_machine` but decided against it or to do it later.

File: pmv-sol/test/idlToTs.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
+import { watch, readdir, readFile, writeFile } from "fs/promises";
```
Comment (freddyaboulton) [inline pmv-sol/test/idlToTs.js:1]: Needed to get the typescript to work in the unit tests. Yanked this from metaplex.

File: pmv-sol/test/nft-candy-machine-types.ts [hunk 8]
```diff
@@ -0,0 +1,14 @@
...[{"code":300,"name":"IncorrectOwner","msg":"Account does not have correct owner!"},{"code":301,"name":"Uninitialized","msg":"Account is not initialized!"},{"code":302,"name":"MintMismatch","msg":"Mint Mismatch!"},{"code":303,"name":"IndexGreaterThanLength","msg":"Index greater than length!"},{"code":304,"name":"ConfigMustHaveAtleastOneEntry","msg":"Config must have atleast one entry!"},{"code":305,"name":"NumericalOverflowError","msg":"Numerical overflow error!"},{"code":306,"name":"TooManyCreators","msg":"Can only provide up to 4 creators to candy machine (because candy machine is one)!"},{"code":307,"name":"UuidMustBeExactly6Length","msg":"Uuid must be exactly of 6 length"},{"code":308,"name":"NotEnoughTokens","msg":"Not enough tokens to pay for this minting"},{"code":309,"name":"NotEnoughSOL","msg":"Not enough SOL to pay for this minting"},{"code":310,"name":"TokenTransferFailed","msg":"Token transfer failed"},{"code":311,"name":"CandyMachineEmpty","msg":"Candy machine is empty!"},{"code":312,"name":"CandyMachineNotLiveYet","msg":"Candy machine is not live yet!"},{"code":313,"name":"ConfigLineMismatch","msg":"Number of config lines must be at least number of items available"}]};
```
Comment (freddyaboulton) [inline pmv-sol/test/nft-candy-machine-types.ts:1]: Need this for typescript in unit tests.

File: pmv-sol/programs/pmv/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+    pub fn mint_nft<'info>(
+        ctx: Context<'_, '_, '_, 'info, MintNFT<'info>>,
+        _bump: u8,
+        token_index: u64,
+    ) -> ProgramResult {
+        let candy_machine = &mut ctx.accounts.candy_machine;
+        let config = &ctx.accounts.config;
+        let claim_status = &mut ctx.accounts.claim_status;
+
+        if let Some(mint) = candy_machine.token_mint {
```
Comment (freddyaboulton) [inline pmv-sol/programs/pmv/src/lib.rs:37]: I can add a 

```rust
if *ctx.accounts.payer.key != candy_machine.authority {
      return Err(ErrorCode::UnApprovedAccess.into());
```

To make sure no one mints but us.

File: pmv-sol/Anchor.toml (hunk 3 above)
Comment (chrishol) [reply pmv-sol/Anchor.toml:1]: First time I've ever seen toml actually used anywhere 🤣

File: pmv-sol/app/app.js (hunk 5 above)
Comment (chrishol) [reply pmv-sol/app/app.js:14]: POST to /claim (or /claim/tokenId) and pass the rest of the params in the request body?

**Review by chrishol — APPROVED** [review]
Nice work! I don't understand any of the Solana stuff, so we may want to get another pair of eyes on that. I think the most important thing is to make sure the contract/endpoint is secure.

File: pmv-sol/programs/pmv/src/lib.rs (hunk 9 above)
Comment (freddyaboulton) [reply pmv-sol/programs/pmv/src/lib.rs:37]: Filed #19 to track this

File: pmv-sol/app/app.js (hunk 5 above)
Comment (freddyaboulton) [reply pmv-sol/app/app.js]: Done! Thanks for the suggestion

**Pull request merged**